pretyping/reductionops.ml

```diff
@@ -1110,6 +1110,37 @@ let sigma_univ_state =
     compare_instances = sigma_compare_instances;
     compare_cumul_instances = sigma_check_inductive_instances; }
 
+let univproblem_compare_sorts env pb s0 s1 uset =
+  let open UnivProblem in
+  match pb with
+  | Reduction.CONV -> UnivProblem.Set.add (ULe (Sorts.univ_of_sort s0, Sorts.univ_of_sort s1)) uset
+  | Reduction.CUMUL -> UnivProblem.Set.add (UEq (Sorts.univ_of_sort s0, Sorts.univ_of_sort s1)) uset
+
+let univproblem_compare_instances ~flex i0 i1 uset =
+  try UnivProblem.enforce_eq_instances_univs flex i0 i1 uset
+  with Evd.UniversesDiffer
+     | Univ.UniverseInconsistency _ ->
+        raise Reduction.NotConvertible
```

Comment:
This is probably the source of the problem.

Reply:
This is essentially a copy-pasting of the same code for sigma_univ_state though...

Reply:
The original actually checks the constraints, so it may raise. This version is just a Set.add and doesn't raise.

Reply:
But we catch this later on, in Unification. Do you mean that it doesn't commute with conversion, so that failing later prevents unfolding needed for unification?

Reply:
Maybe not that related
It's still a nnsense try/catch

Reply:
@ppedrot I believe it is not commuting due to the first-order unification heuristic: conversion might backtrack if some universe unification fails with UniversesDiffer, and this might happen because some constraints that were there are now postponed. If it no longer backtracks then you've changed the semantics.